src/tests/ftest/launch.py

```diff
@@ -128,11 +128,40 @@ def set_test_environment():
     usr_sbin = os.path.join("/usr/", "sbin")
     path = os.environ.get("PATH")
 
+    # Get the default interface to use if OFI_INTERFACE is not set
+    #   - Only include interfaces that are active/up (operstate)
+    #   - Exclude the loopback device (lo)
+    #   - Only include the first active interface for each interface type
+    #   - Sorting the ARP protocol HW identifiers (type) to select ib over eth
+    print("Detecting network devices")
+    available_interfaces = {}
+    net_path = os.path.join(os.path.sep, "sys", "class", "net")
+    for device in sorted([dev for dev in os.listdir(net_path) if dev != "lo"]):
+        with open(os.path.join(net_path, device, "operstate"), "r") as buffer:
+            state = buffer.read().strip()
+        with open(os.path.join(net_path, device, "type"), "r") as buffer:
+            dtype = int(buffer.read().strip())
+        print(
+            "  - {0:<5} (type: {1:<2} state: {2})".format(
+                device, dtype, state))
+        if state.lower() == "up" and dtype not in available_interfaces:
+            available_interfaces[dtype] = device
+    try:
+        interface = available_interfaces[sorted(available_interfaces)[-1]]
```

Comment:
Not a review, so feel free to rebase if you want even though I've commented but what's the sort based on here?  Is it just that _ib0_ is sorted alphabetically after _eth*_?  What if there is an _ib1_?  Won't than end up being the chosen interface even though we really want _ib0_.

An ideal sort (not to imply doing that here and now in this PR) would be to use interface speed.  Not sure how to get that programmatically without some assumptions and constants.

But for now at least, just wanted to flag the possibility of getting _ib1_ when you really want _ib0_.

Reply:
The sort is based upon the keys of the `available_interfaces` dictionary.  The keys are the integer value read from the `/sys/class/net/<device>/type` file which correspond to the hardware identifier enumerations in https://github.com/torvalds/linux/blob/master/include/uapi/linux/if_arp.h.

Prior to this line of code, on line 139, the names of the devices were used to sort the listing from `/sys/class/net/` (excluding `lo`).  The result of this sort will yield a `available_interfaces` dictionary with one key per device type (integer) and a value the consists of the first device encountered of that type.  In other words if there are two ARPHRD_ETHER (1) devices - eth0, eth1 - and two ARPHRD_INFINIBAND (32) devices - ib0 and ib1 - then `available_interfaces = {1: 'eth0', 32: 'ib0'}`.

So we will only get `ib1` instead of `ib0` if 'ib1' is up and 'ib0' is down.

Instead of using the type I could use `/sys/class/net/<device>/speed` value to index the `available_interfaces` dictionary:
```
$ for dev in eth0 eth1 ib0; do echo -n "${dev}: "; cat /sys/class/net/${dev}/speed; done
eth0: 1000
eth1: -1
ib0: 56000
```
(eth1's speed is -1 because its state is down)

Reply:
Sample logging with speed sort:
```
Detecting network devices
  - eth0  (speed:   1000 state: up)
  - eth1  (speed:     -1 state: down)
  - ib0   (speed:  56000 state: up)
Using ib0 as the default interface from {1000: 'eth0', 56000: 'ib0'}
```

Reply:
Interesting that speed is even available.  56000 is an *interesting* speed for `ib0`, though it seems to serve our purposes.  Probably in the above, `eth1` should sort before `eth0` so that `eth0` is the chosen device when there is no `ib0`.

> value the consists of the first device encountered of that type. In other words if there are two ARPHRD_ETHER (1) devices - eth0, eth1 - and two ARPHRD_INFINIBAND (32) devices - ib0 and ib1 - then available_interfaces = {1: 'eth0', 32: 'ib0'}.

Thanks for the clarification.  Suggests that the existing code is sufficient even when there is an `ib1`.  In that case I will leave it up to you to decide if you want to go with the existing logic or use speed as a selector, or something else even.  I guess a balance between "over-engineering" and getting something landed soon that reasonably uses `ib0` and doesn't break as soon as the environment changes (i.e more IB interfaces become available).

Reply:
On wolf-79:
$ cat /sys/class/net/ib0/speed
100000

Reply:
Yes, we saw a similar speed in the Functional HW run (https://build.hpdd.intel.com/blue/rest/organizations/jenkins/pipelines/daos-stack/pipelines/daos/branches/PR-1520/runs/11/nodes/475/steps/605/log/?start=0):
```
[2019-12-09T22:12:33.167Z] Detecting network devices
[2019-12-09T22:12:33.167Z]   - eth0  (speed:   1000 state: up)
[2019-12-09T22:12:33.167Z]   - eth1  (speed:     -1 state: down)
[2019-12-09T22:12:33.167Z]   - ib0   (speed: 100000 state: up)
[2019-12-09T22:12:33.167Z] Using ib0 as the default interface from {1000: 'eth0', 100000: 'ib0'}
```